### PR TITLE
Refactor project source code

### PR DESCRIPTION
**Pull request by jackyhuang85**
* remove `verbose message`
* remove `victim cache`
* add `InstructionLoader` to parse tracefile
* refactor functions

### CODE REVIEW

File: src/loader.cpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    addr_raw_t addr = (addr_raw_t)strtoul(line + 2, nullptr, INST_ADDR_BASE);
+
+    switch (line[0]) {
+    case 'l':
+        // TODO: add statistics
+        op = I_LOAD;
+        break;
+    case 's':
+        // TODO: add statistics
+        op = I_STORE;
```
Comment (jerryzj) [inline src/loader.cpp]: Codacy Issue found: [Variable 'op' is assigned a value that is never used.](https://app.codacy.com/app/jerry0715no14/Cache_sim/pullRequest?prid=4720103)

**Comment by jerryzj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 7
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/jerry0715no14/Cache_sim/pullRequest?prid=4720103&bid=15760136)

File: src/loader.hpp [hunk 2]
```diff
@@ -0,0 +1,22 @@
+#ifndef _LOADER_HPP_
+#define _LOADER_HPP_
+
+#include "datatype.hpp"
+#include <fstream>
+#include <iostream>
+
+class InstructionLoader {
+  public:
+    InstructionLoader(const std::string &filename);
```
Comment (jerryzj) [inline src/loader.hpp]: Codacy Issue found: [Class 'InstructionLoader' has a constructor with 1 argument that is not explicit.](https://app.codacy.com/app/jerry0715no14/Cache_sim/pullRequest?prid=4720103)

File: src/loader.cpp [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+void InstructionLoader::LoadTraceFile(const std::string &filename) {
+    in_file = std::make_unique<std::ifstream>(filename.c_str(), std::ios::in);
+    if (in_file->fail()) {
+        std::cerr << "Open trace file error" << std::endl;
+        exit(-1);
+    }
+}
+
+inst_t InstructionLoader::GetNextInst() {
+    int LENGTH_OF_INST_LINE = 13;
```
Comment (jerryzj) [inline src/loader.cpp]: Codacy Issue found: [The scope of the variable 'LENGTH_OF_INST_LINE' can be reduced.](https://app.codacy.com/app/jerry0715no14/Cache_sim/pullRequest?prid=4720103)

File: src/cache.cpp [hunk 4]
```diff
@@ -201,105 +218,75 @@ std::bitset<32> Cache::_CvtToAddr(const ulint block_set) {
...
-                    _poten_victim |= std::bitset<32>(0xffffffff);
-                    break;
+            if (!_IfBlockAvailable()) {
+                _has_evicted = true;
+                if (_cache_setting.replacement_policy == RANDOM) {
+                    int _ = GetBlockByRandom();
+                    _++;
+                } else if (_cache_setting.replacement_policy == LRU) {
+                    int _ = GetBlockByLRU();
+                    _++;
```
Comment (jerryzj) [inline src/cache.cpp]: Codacy Issue found: [Variable '_' is assigned a value that is never used.](https://app.codacy.com/app/jerry0715no14/Cache_sim/pullRequest?prid=4720103)

File: src/datatype.hpp [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+inline addr_t Cvt2AddrBits(addr_raw_t raw_addr) {
+    return std::bitset<32>(raw_addr);
+}
+
+enum INST_OP { I_LOAD, I_STORE, I_NONE };
+
+using inst_t = struct {
+    INST_OP op;
+    addr_raw_t addr_raw;
```
Comment (jerryzj) [inline src/datatype.hpp]: Codacy Issue found: [struct member 'inst_t::addr_raw' is never used.](https://app.codacy.com/app/jerry0715no14/Cache_sim/pullRequest?prid=4720103)

File: src/loader.cpp [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    case 's':
+        // TODO: add statistics
+        op = I_STORE;
+        break;
+    case '\0':
+        // TODO: add statistics
+        op = I_NONE;
+        break;
+    default:
+        op = I_NONE;
```
Comment (jerryzj) [inline src/loader.cpp]: Codacy Issue found: [Variable 'op' is assigned a value that is never used.](https://app.codacy.com/app/jerry0715no14/Cache_sim/pullRequest?prid=4720103)

File: src/loader.cpp [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+inst_t InstructionLoader::_ParseLineToInst(const char *line) {
+    int INST_ADDR_BASE = 16;
+    INST_OP op;
+    addr_raw_t addr = (addr_raw_t)strtoul(line + 2, nullptr, INST_ADDR_BASE);
+
+    switch (line[0]) {
+    case 'l':
+        // TODO: add statistics
+        op = I_LOAD;
```
Comment (jerryzj) [inline src/loader.cpp]: Codacy Issue found: [Variable 'op' is assigned a value that is never used.](https://app.codacy.com/app/jerry0715no14/Cache_sim/pullRequest?prid=4720103)

**Comment by jerryzj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/jerry0715no14/Cache_sim/pullRequest?prid=4720103&bid=15760136)

File: src/cache.cpp (hunk 4 above)
Comment (jerryzj) [inline src/cache.cpp]: Codacy Issue found: [Variable '_' is assigned a value that is never used.](https://app.codacy.com/app/jerry0715no14/Cache_sim/pullRequest?prid=4720103)

File: src/loader.cpp [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        // TODO: add statistics
+        return inst_t(I_STORE, addr);
+        break;
+    case '\0':
+        // TODO: add statistics
+        return inst_t(I_NONE, addr);
+        break;
+    default:
+        return inst_t(I_NONE, addr);
+        std::cerr << "Undefined instruction type." << std::endl;
```
Comment (jerryzj) [inline src/loader.cpp]: Codacy Issue found: [Statements following return, break, continue, goto or throw will never be executed.](https://app.codacy.com/app/jerry0715no14/Cache_sim/pullRequest?prid=4720103)

File: src/datatype.hpp [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+inline addr_t Cvt2AddrBits(addr_raw_t raw_addr) {
+    return std::bitset<32>(raw_addr);
+}
+
+enum INST_OP { I_LOAD, I_STORE, I_NONE };
+
+using inst_t = struct inst_s_t {
+    INST_OP op;
+    addr_raw_t addr_raw;
+    inst_s_t() : op(op), addr_raw(addr_raw) {}
```
Comment (jerryzj) [inline src/datatype.hpp]: Codacy Issue found: [Member variable 'addr_raw' is initialized by itself.](https://app.codacy.com/app/jerry0715no14/Cache_sim/pullRequest?prid=4720103)

File: src/loader.cpp (hunk 8 above)
Comment (jerryzj) [reply src/loader.cpp]: @jackyhuang85  in default branch, error message should be printed before return

File: src/datatype.hpp (hunk 9 above)
Comment (jerryzj) [reply src/datatype.hpp]: @jackyhuang85  Is there any reason to declare a struct and use `using` to give it another type name?

**Review by jerryzj — APPROVED** [review]

File: src/cache.cpp [hunk 10]
```diff
@@ -201,32 +201,32 @@ std::bitset<32> Cache::_CvtToAddr(const ulint block_set) {
     addr.reset();
     std::bitset<32> index(block_set);
 
+    auto a_tag = [_c = this](std::bitset<32> &dst_addr, const ulint block_set) {
```
Comment (jerryzj) [inline src/cache.cpp]: I think `a_tag` is meaningless

File: src/loader.hpp [hunk 11]
```diff
@@ -0,0 +1,22 @@
+#ifndef _LOADER_HPP_
+#define _LOADER_HPP_
+
+#include "datatype.hpp"
+#include <fstream>
+#include <iostream>
```
Comment (jerryzj) [inline src/loader.hpp]: `#include <fstream>` and `#include <iostream>` should be added in `datatype.hpp`

File: src/simulator.cpp [hunk 12]
```diff
@@ -97,69 +98,57 @@ void Simulator::ReadConfig() {
...
 
-    // Preprocess the following memory address
-    std::bitset<32> poten_victim_addr = this->main_cache->Ready(addr);
+    std::bitset<32> poten_victim_addr = this->main_cache->Ready(addr_b);
 
     // Handle the address
     if (is_load) {
         ++_counter.access;
         ++_counter.load;
+<<<<<<< HEAD
```
Comment (jerryzj) [inline src/simulator.cpp]: Brach history leftover.....

File: src/simulator.cpp [hunk 13]
```diff
@@ -97,69 +98,57 @@ void Simulator::ReadConfig() {
...
-    default:
-        std::cerr << "Undefined instruction type." << std::endl;
-        std::cerr << "Error line: " << trace_line << std::endl;
-        return false;
     }
 
-    // Parse the address from the instruction
-    auto temp = strtoul(trace_line + 2, nullptr, INST_ADDR_BASE);
-    std::bitset<32> addr(temp);
+    addr_t addr_b = Cvt2AddrBits(inst.addr_raw);
```
Comment (jerryzj) [inline src/simulator.cpp]: What is addr_b?

**Review by jerryzj — COMMENTED** [review]
@jackyhuang85  You have to make sure that the template can be compiled successfully.

File: src/simulator.cpp (hunk 12 above)
Comment (jackyhuang85) [reply src/simulator.cpp]: Interesting.. I don't see any "HEAD" in my source code

**Comment by jackyhuang85** [comment]
@jerryzj 
Checked ! 
Can be compiled successfully

**Comment by jerryzj** [comment]
> @jerryzj
> Checked !
> Can be compiled successfully

Can't make it on my desktop.
Fedora 31 kernel 5.3.16
clang 9.0.0 with CMake 3.16.1

image

**Comment by jerryzj** [comment]
> @jerryzj
> Checked !
> Can be compiled successfully

Please note that Apple clang has its own branch, it contains many extra protections while the official LLVM doesn't have.

**Pull request merged**